identity-and-access-management/key-manager/how-to/use-tink-go.mdx

```diff
@@ -0,0 +1,310 @@
+See the full list of supported algorithms and ciphers in the
+[Tink Go reference](https://pkg.go.dev/github.com/tink-crypto/tink-go/v2/aead)
+if you need something else. Stick with AES256-GCM if you are unsure.
+</Message>
+
+
+The DEK is now stored in `/path/to/encrypted_dek.tink`. When the application
+needs it to do meaningful work, the keyset can be read back as follows:
+
+```go
+// Load, deserialize, and decrypt the DEK with the remote KEK
+f, _ := os.Open("/path/to/encrypted_dek.tink")
+defer f.Close()
+r := keyset.NewBinaryReader(f)
+handle, _ := keyset.ReadWithAssociatedData(r, kekAEAD, associatedData)
+
+// Use the DEK to encrypt some data
+primitive, _ := aead.New(handle)
+
+secret1, _ := primitive.Encrypt([]byte("This is a secret"), dataAssociatedData1)
+/*
+ *   Store secret1 somewhere
+ */
+
+secret2, _ := primitive.Encrypt([]byte("This is another secret"), dataAssociatedData2)
+/*
+ *   Store secret2 somewhere
+ */
+```
+
+Tink only provides methods working on types that comply with `Reader` and
+`Writer` interfaces. If you need to write and read Tink keysets as direct bytes,
+you can use `bytes.Buffer`.
+
+```go
+buf := new(bytes.Buffer)
+w := keyset.NewBinaryWriter(buf)
+handle.WriteWithAssociatedData(w, kekAEAD, associatedData)
+encDEK := buf.Bytes()  // encrypted DEK in Tink wire format
+```
+
+You can then store the bytes of the encrypted DEK into a database for example,
+alongside the encrypted data it protects. (base64-encoded in the following example.)
+
+```console
+SELECT id, enc_data, enc_dek FROM sensible_stuff;
+
+ id  | enc_data                   | enc_dek
+-----|----------------------------|-------------------
+ 42  | "7NXIqRms0+TiKj+V0gv1s..." | "vIiYBeypb7Yk..."
+ 43  | "7X8v0GVrXWwL/ckzfRms0..." | "vIiYBeypb7Yk..."
+...
+...
+...
+```
+
+
+## Associated Data
+
+Associated data (AD) is not
+encrypted, but it is authenticated. That is, it must be the same when
+supplied to both `Encrypt` and `Decrypt`, otherwise the decryption fails.
+
+This is useful to prevent reading the wrong data in the wrong context. In
+the previous example, the data in both rows 42 and 42 is protected by the
+same DEK. If we swapped the data, an application would be able to decrypt
+the data from another row with success. But, by providing the intended ID as the
+associated data, the decryption would fail.
+
+Here is a way we could encrypt the data with AD before inserting it:
+
+```go
+handle, _ := keyset.ReadWithAssociatedData(r, kekAEAD, dekAD)
+primitive, _ := aead.New(handle)  // Same DEK for the two payloads
+
+secret1, _ := primitive.Encrypt([]byte("This is a secret"), []byte("id42"))
+// Insert secret1 into row 42
+
+secret2, _ := primitive.Encrypt([]byte("This is another secret"), []byte("id43"))
+// Insert secret2 into row 43
+```
+
+You should preferably use AD in general. AD does not need to be stored, as
+it can be infered from the context at decryption time.
+
+It is also possible to use a unique DEK for each payload.
+
```

Comment:
```suggestion
---
meta:
  title: Using Tink with the Scaleway Key Manager
  description: Discover essential concepts of Scaleway Key Manager and how to use it effectively with the Tink Go library
content:
  h1: Using Tink with the Scaleway Key Manager
  paragraph: Discover essential concepts of Scaleway Key Manager, including Key Encryption Keys and Data Encryption Keys
tags: key-manager encryption kms key-management-service cryptographic
dates:
  validation: 2024-09-20
---

This documentation page provides information on Key Manager Key Encryption Keys (KEKs) and Data Encryption Keys (DEKs), and how to use them with the Tink Go library.

Tink is an open-source tool that helps you safely perform cryptographic operations, like encrypting data and managing encryption keys. It works well with different platforms and follows best practices for security. Tink also supports using remote keys right away.

We suggest using Tink when working with Scaleway's Key Manager, especially with Go or Python.

<Macro id="requirements" />

- A Scaleway account logged into the [console](https://console.scaleway.com)
- [Owner](/identity-and-access-management/iam/concepts/#owner) status or [IAM permissions](/identity-and-access-management/iam/concepts/#permission) allowing you to perform actions in the intended Organization
- Created and enabled a kata encryption key (DEK) in your Key Manager
- A working Go environment
- Installed the [Scaleway Go SDK](https://github.com/scaleway/scaleway-sdk-go) with [valid credentials](/developer-tools/scaleway-sdk/go-sdk/)


## Installing the necessary tools

Tink is a tool that helps you perform encryption (securing data) and manage encryption keys. It can work with various key management services (KMS), including Scaleway's Key Manager.
To use Tink with Scaleway Key Manager, you need to install dependencies that let Tink interact with Key Manager.

1. Open a terminal and run the following commands:

    ```shell
    # Install Tink for Go
    go get -u github.com/tink-crypto/tink-go/v2

    # Install the Scaleway Tink extension
    go get -u github.com/scaleway/tink-go-scwkms
    ```
2. Retrieve the ID (UUIDv4 format) of your Key Manager's key (KEK).
3. Copy the following template and paste it into a `.go` file:

    ```go
    import (
        "github.com/scaleway/scaleway-sdk-go/scw" //  Library that helps your Go program interact with Scaleway
        "github.com/tink-crypto/tink-go/v2/aead" // Tink's encryption tools
        "github.com/scaleway/tink-go-scwkms/integration/scwkms" // Scaleway's Tink extension
    )

    const region = "<REGION>"                              // Replace the placeholder with the region where your key is created
    const keyID = "7f967268-bebb-43b0-9fe2-e13bd23bf421" // Replace the placeholder with the unique ID of your key encryption key

    keyURIPrefix := "scw-kms://regions/" + region + "/keys/"
    keyURI := keyURIPrefix + keyID

    // Set up a connection to Scaleway
    config, _ := scw.LoadConfig() // Loads your Scaleway configuration
    profile, _ := config.GetActiveProfile() // Gets the active profile (your account settings)

    // Set up the connection to use your key in Key Manager
    kms, _ := scwkms.NewClientWithOptions(
        keyURIPrefix,
        scw.WithProfile(profile), // Uses your account profile
        scw.WithEnv(), // Uses environment settings
    )
    // Prepare the key for encryption and decryption
    kekAEAD, _ := kms.GetAEAD(keyURI)
    ```
4. Replace the placeholder values with your own.

    <Message type="note">
     The `kekAEAD` object represents the key in Scaleway’s Key Manager. It allows you to encrypt payloads (secure) and decrypt ciphertexts (unlock).
    </Message>

## Encrypting and decrypting data

Paste the following code into a `.go` file. This file contains the data we will encrypt (`Hello, World!`), and the code to encrypt and decrypt it.
    ```go
    associatedData := []byte("") // Read the ## Associated data section for more information
    secretData := []byte("Hello, World!") // Data we want to encrypt

    ciphertext, _ := kekAEAD.Encrypt(secretData, associatedData) // Encrypt the data
    fmt.Println(ciphertext) // Print the encrypted data

    plaintext, _ := kekAEAD.Decrypt([]byte(ciphertext), associatedData)
    fmt.Println(string(plaintext)) // Output: "Hello, World!"
    ```

<Message type="important">
 While the code shown above functions as intended, this is not a recommended pattern and the following limitations apply:
  - It is slow: since the key resides on Scaleway Key Manager, each encryption or decryption operation translates into a remote API call.
  - The payload to encrypt is limited in size: Key Manager only allows up to 64 KiB. As a result, you will not be able to encrypt larger payloads with `kekAEAD`.
  - You cannot choose the cipher and the algorithm that suit your use-case, Key Manager handles that on your behalf.
</Message>

## Understanding the difference between KEKs and DEKs

**Key encryption key (KEK)**: This is a key (`kekAEAD`) used to encrypt other keys (DEKs, which protect your data), not the actual data. It stays secure in Scaleway Key Manager.

**Data Encryption Key (DEK)**: This key is used to encrypt your actual data. Key Manager does not store or manage your DEK. A DEK is usually stored alongside the data it protects, and it is **always** stored encrypted (by the  KEK).

The KEK secures the DEK, and the DEK secures your data. This double-layer approach improves security.

<Message type="important">
 - Your application is responsible for storing the encrypted DEKs alongside the encrypted data they protect.
 - DEKs must **never** reside in plaintext. They must be encrypted by the remote KEK before being stored.
</Message>

## Using Tink keysets as DEKs

Tink does not just handle single keys, it also manages groups of keys called **keysets**, a set of related keys kept together with their metadata. We use the terms `key` and `keyset` interchangeably since we work with keysets containing only one key here. Tink can generate keys to be used with the desired algorithm and cipher.

1. Run the following code to create a DEK using Tink and store it encrypted with your KEK, using the `AES256-GCM` algorithm.

    ```go
    import "github.com/tink-crypto/tink-go/v2/keyset" // Import the package that provides functionalities to create, manage, and store keysets in Tink

    /* ... */
    /* ... */

    handle, _ := keyset.NewHandle(aead.AES256GCMKeyTemplate())  // Generate an AES256-GCM key
    associatedData := []byte("")

    f, _ := os.OpenFile("/path/to/encrypted_dek.tink", os.O_RDWR|os.O_CREATE, 0644)
    defer f.Close()
    w := keyset.NewBinaryWriter(f)

    // Serialize the keyset and encrypt it with the remote KEK,
    handle.WriteWithAssociatedData(w, kekAEAD, associatedData)
    ```


    <Message type="note">
     Check out the full list of supported algorithms and ciphers in the [Tink Go reference](https://pkg.go.dev/github.com/tink-crypto/tink-go/v2/aead) if you need to use another algorithm or cipher. We recommend that you stick with the `AES256-GCM` algorithm if you are unsure.
    </Message>


     Your encrypted DEK is now stored in `/path/to/encrypted_dek.tink`.

2. Run the following code to read and use your encrypted DEK:

    ```go
    // Load, deserialize, and decrypt the DEK with the remote KEK
    f, _ := os.Open("/path/to/encrypted_dek.tink")
    defer f.Close()
    r := keyset.NewBinaryReader(f)
    handle, _ := keyset.ReadWithAssociatedData(r, kekAEAD, associatedData)

    // Use the DEK represented by "primitive" to encrypt data. The primitive created from the DEK is used to encrypt a string ("This is a secret") into a ciphertext (secret1)
    primitive, _ := aead.New(handle)

    secret1, _ := primitive.Encrypt([]byte("This is a secret"), dataAssociatedData1)
    /*
    *   Store secret1 somewhere
    */

    secret2, _ := primitive.Encrypt([]byte("This is another secret"), dataAssociatedData2)
    /*
    *   Store secret2 somewhere
    */
    ```
    <Message type="note">
     Tink only provides methods that work on types that comply with `Reader` and `Writer` interfaces. If you need to write and read Tink keysets as direct bytes, you can use `bytes.Buffer`, which is an in-memory buffer used to hold the encrypted keyset. This allows you to serialize and deserialize keysets directly as byte arrays instead of using files.
    </Message>

3. Run the following command to store the DEK in memory as bytes:

    ```go
    buf := new(bytes.Buffer)
    w := keyset.NewBinaryWriter(buf)
    handle.WriteWithAssociatedData(w, kekAEAD, associatedData)
    encDEK := buf.Bytes()  // encrypted DEK in Tink wire format
    ```

You can then store the bytes of the encrypted DEK in a database for example, with the encrypted data it protects. For example, the encrypted data (enc_data) and the encrypted DEK (enc_dek) might be stored together in a row in a database (base64-encoded in the following example):

    ```console
    SELECT id, enc_data, enc_dek FROM sensible_stuff;

    id  | enc_data                   | enc_dek
    -----|----------------------------|-------------------
    42  | "7NXIqRms0+TiKj+V0gv1s..." | "vIiYBeypb7Yk..."
    43  | "7X8v0GVrXWwL/ckzfRms0..." | "vIiYBeypb7Yk..."
    ...
    ...
    ...
    ```


## Associated Data

Associated Data (AD) is not encrypted, but it is authenticated. It must be the same when you encrypt and decrypt data, otherwise the decryption fails. This is useful to prevent reading the wrong data in the wrong context. In the table above, the data in both rows 42 and 43 is protected by the same DEK. If we swapped the data, an application would be able to decrypt the data from another row. But, by providing the intended ID as the associated data, the decryption would fail.

### Encrypt data with AD before inserting it

Run the following command to encrypt your data with `Associated Data`. In the example below, associated data like `id42` and `id43` is used to ensure that data from row 42 cannot be decrypted in the context of row 43.

    ```go
    handle, _ := keyset.ReadWithAssociatedData(r, kekAEAD, dekAD)
    primitive, _ := aead.New(handle)  // Same DEK for the two payloads

    secret1, _ := primitive.Encrypt([]byte("This is a secret"), []byte("id42"))
    // Insert secret1 into row 42

    secret2, _ := primitive.Encrypt([]byte("This is another secret"), []byte("id43"))
    // Insert secret2 into row 43
    ```

Associated Data does not need to be stored, as it can be inferred from the context at decryption time. It is also possible to use a unique DEK for each payload. We recommend using Associated Data.
```

Reply:
I'm still reviewing the rest of the content :)